sys/drv/bcm2835_emmc.c

```diff
@@ -0,0 +1,535 @@
+      break;
+    case EMMC_PROP_RW_RESP_HI:
+      *var = (uint64_t)b_in(emmc, BCMEMMC_RESP2) |
+             (uint64_t)b_in(emmc, BCMEMMC_RESP3) << 32;
+      break;
+    case EMMC_PROP_RW_BUSWIDTH:
+      *var = bcmemmc_get_bus_width(state);
+      break;
+    case EMMC_PROP_RW_RCA:
+      *var = state->rca;
+      break;
+    default:
+      return EINVAL;
+  }
+
+  return 0;
+}
+
+static int bcmemmc_set_prop(device_t *cdev, uint32_t id, uint64_t var) {
+  bcmemmc_state_t *state = (bcmemmc_state_t *)cdev->parent->state;
+  resource_t *emmc = state->emmc;
+
+  uint32_t reg = 0;
+  switch (id) {
+    case EMMC_PROP_RW_BLKCNT:
+      if (var >= 65536)
+        return EINVAL;
+      reg = b_in(emmc, BCMEMMC_BLKSIZECNT) & ~BSC_BLKCNT;
+      b_out(emmc, BCMEMMC_BLKSIZECNT, reg | (var << BSC_BLKCNT_SHIFT));
+      break;
+    case EMMC_PROP_RW_BLKSIZE:
+      if (var >= 1024)
+        return EINVAL;
+      reg = b_in(emmc, BCMEMMC_BLKSIZECNT) & ~BSC_BLKSIZE;
+      b_out(emmc, BCMEMMC_BLKSIZECNT, reg | var);
+      break;
+    case EMMC_PROP_RW_RESP_LOW:
+      b_out(emmc, BCMEMMC_RESP0, (uint32_t)var);
+      b_out(emmc, BCMEMMC_RESP1, (uint32_t)(var >> 32));
+      break;
+    case EMMC_PROP_RW_RESP_HI:
+      b_out(emmc, BCMEMMC_RESP2, (uint32_t)var);
+      b_out(emmc, BCMEMMC_RESP3, (uint32_t)(var >> 32));
+      break;
+    case EMMC_PROP_RW_CLOCK_FREQ:
+      return bcmemmc_set_clk_freq(cdev->parent, var);
+    case EMMC_PROP_RW_BUSWIDTH:
+      return bcmemmc_set_bus_width(state, var);
+    case EMMC_PROP_RW_RCA:
+      state->rca = var;
+      break;
+    default:
+      return EINVAL;
+  }
+
+  return 0;
+}
+
+static int bcmemmc_send_cmd(device_t *cdev, emmc_cmd_t cmd, uint32_t arg,
+                            emmc_resp_t *resp) {
+  device_t *dev = cdev->parent;
+  bcmemmc_state_t *state = (bcmemmc_state_t *)dev->state;
+  resource_t *emmc = state->emmc;
+  uint32_t code = bcmemmc_encode_cmd(cmd);
+
+  if (cmd.flags & EMMC_F_APP)
+    bcmemmc_send_cmd(cdev, EMMC_CMD(APP_CMD), state->rca << 16, NULL);
+
+  b_out(emmc, BCMEMMC_ARG1, arg);
+  b_out(emmc, BCMEMMC_CMDTM, code);
+
+  int error = bcmemmc_intr_wait(dev, INT_CMD_DONE);
+  if (error) {
+    klog("ERROR: failed to send EMMC command %p", code);
+    return error;
+  }
+
+  if (resp) {
+    resp->r[0] = b_in(emmc, BCMEMMC_RESP0);
+    resp->r[1] = b_in(emmc, BCMEMMC_RESP1);
+    resp->r[2] = b_in(emmc, BCMEMMC_RESP2);
+    resp->r[3] = b_in(emmc, BCMEMMC_RESP3);
+  }
+
+  return 0;
+}
+
+static int bcmemmc_read(device_t *cdev, void *buf, size_t len, size_t *donep) {
+  device_t *emmcdev = cdev->parent;
+  bcmemmc_state_t *state = (bcmemmc_state_t *)emmcdev->state;
+  resource_t *emmc = state->emmc;
+  uint32_t *data = buf;
+
+  assert(is_aligned(len, sizeof(uint32_t)));
+
+  /* TODO(morhcore): should use DMA transfer if possible. */
+  for (size_t i = 0; i < len / sizeof(uint32_t); i++)
+    data[i] = b_in(emmc, BCMEMMC_DATA);
```

Comment:
Should not `len` be equal to currently selected block size?

Reply:
No. The read/write operations can be performed in other scenarios than block transfers. See `sd_init` in. #1135
There's 64 byte read just after issuing `SD_CMD_SEND_SCR` command.